src/olmo_core/optim/adamw.py

```diff
@@ -115,6 +166,61 @@ def step(self, closure=None) -> None:
                 )
 
 
+class SkipStepAdamWForeach(SkipStepAdamW):
+    """
+    A "skip step" version of AdamW that updates parameters using PyTorch
+    multi‐tensor (*foreach*) kernels.  This executes the heavy math in a handful
+    of fused CUDA kernels instead of one kernel per parameter making it much faster.
+    """
+
+    @torch.no_grad()
+    def step(self, closure=None) -> None:
+        if closure is not None:
+            with torch.enable_grad():
+                closure()
+
+        step_factor = self.get_step_factor()
+        self._step_skipped = 1 - step_factor
+        for group in self.param_groups:
+            params_with_grad: list[torch.Tensor] = []
+            grads: list[torch.Tensor] = []
+            exp_avgs: list[torch.Tensor] = []
+            exp_avg_sqs: list[torch.Tensor] = []
+            steps_list = []  # create list outside loops
+
+            for p in group["params"]:
+                if p.grad is None:
+                    continue
+
+                state = self.state[p]
+                if len(state) == 0:
+                    state["step"] = torch.tensor(0.0, dtype=torch.float32, device=p.device)
```

Comment:
I'm 99% sure there is a bug in SkipStepAdamW (replicated here) where state["step"] is never updated. Is there a paper or a doc that describes the SkipStepOptimizer?